homeassistant/components/vilfo/__init__.py

```diff
@@ -0,0 +1,114 @@
+"""The Vilfo Router integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+import vilfo
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.util import Throttle, dt as dt_util
+
+from .const import DOMAIN
+
+PLATFORMS = ["sensor"]
+
+DEFAULT_SCAN_INTERVAL = timedelta(seconds=30)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Vilfo Router component."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Vilfo Router from a config entry."""
+    host = entry.data[CONF_HOST]
+    access_token = entry.data[CONF_ACCESS_TOKEN]
+
+    vilfo_router = VilfoRouterData(host, access_token)
+
+    await vilfo_router.async_update()
+
+    if not vilfo_router.available:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = vilfo_router
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class VilfoRouterData:
+    """Define an object to hold sensor data."""
+
+    def __init__(self, host, access_token):
+        """Initialize."""
+        self._vilfo = vilfo.Client(host, access_token)
+        self.host = host
+        self.available = False
+        self.firmware = None
+        self.mac_address = self._vilfo.mac
+        self.data = {}
+        self.unavailable_logged = False
+
+    @property
+    def unique_id(self):
+        """Get the unique_id for the Vilfo Router."""
+        if self.mac_address:
+            return self.mac_address
+        return self.host
```

Comment:
I can have a MAC address, but since the resolution of it's MAC address (in the `vilfo-api-client`) relies on `get-mac` (https://pypi.org/project/get-mac/) there might be situations where it's not possible to resolve the router MAC. Hence, I thought it would be good to use it if available, but otherwise don't rely on it. But I might have thought wrong 😄

Reply:
An IP address is not allowed as unique id as it may change:
https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

It's ok to use the mac address when that's available, otherwise fall back to `None`.

Reply:
Ok :+1:

As the variable name implies, it can also be a hostname (`admin.vilfo.com` being the default for such a router). I guess a hostname would be fine to use as unique ID? :smile:

Reply:
If it doesn't change a hostname may be ok.

Reply:
@MartinHjelmare I've changed it to only use the `host` as `unique_id` when the `host` is set to `admin.vilfo.com`, which is the default one and that isn't possible to be modified (at least not supported by Vilfo). I've pushed these changes to update the PR.